src/sql/src/names.rs

```diff
@@ -806,30 +806,71 @@ impl<'a> Fold<Raw, Aug> for NameResolver<'a> {
     fn fold_query(&mut self, q: Query<Raw>) -> Query<Aug> {
         // Retain the old values of various CTE names so that we can restore them after we're done
         // planning this SELECT.
-        let mut old_cte_values = Vec::new();
-        // A single WITH block cannot use the same name multiple times.
-        let mut used_names = HashSet::new();
-        let mut ctes = Vec::new();
-        for cte in q.ctes {
-            let cte_name = normalize::ident(cte.alias.name.clone());
-
-            if used_names.contains(&cte_name) {
-                self.status = Err(sql_err!(
-                    "WITH query name \"{}\" specified more than once",
-                    cte_name
-                ));
+        let mut shadowed_cte_ids = Vec::new();
+
+        // A reused identifier indicates a reused name.
+        use itertools::Itertools;
+        if let Some(ident) = q.ctes.bound_identifiers().duplicates().next() {
+            self.status = Err(sql_err!(
+                "WITH query name \"{}\" specified more than once",
+                normalize::ident_ref(ident),
+            ));
+        }
+
+        let ctes: CteBlock<Aug> = match q.ctes {
+            CteBlock::Simple(ctes) => {
+                let mut result_ctes = Vec::<Cte<Aug>>::new();
+
+                let initial_id = self.ctes.len();
+
+                for (offset, cte) in ctes.into_iter().enumerate() {
+                    let cte_name = normalize::ident(cte.alias.name.clone());
+                    let local_id = LocalId::new((initial_id + offset) as u64);
```

Comment:
`mz_ore::cast::cast_from` allows you to cast from `usize` to `u64` without using `as` directly, if you prefer to use that.

Reply:
All done, thanks!